lib/puppet/http/session.rb

```diff
@@ -38,4 +41,17 @@ def route_to(name, url: nil, ssl_context: nil)
   def add_exception(exception)
     @resolution_exceptions << exception
   end
+
+  def process_response(response)
+    version = response[Puppet::HTTP::HEADER_PUPPET_VERSION]
+    if version
+      site = Puppet::Network::HTTP::Site.from_uri(response.url)
+      @server_versions[site] = SemanticPuppet::Version.parse(version)
+    end
+  end
+
+  def server_version(url)
+    site = Puppet::Network::HTTP::Site.from_uri(url)
+    @server_versions[site] || SemanticPuppet::Version.parse("4.0.0")
```

Comment:
So this defaults to returning 4.0.0 if we don't already have an entry for `site` in `@server_versions`, right?

Reply:
yeah. I wanted to make sure this was less than 5.0, so as that's the case where json serialization isn't supported